dry-run: Create Options with dryRun for POST/PUT/PATCH

apelisse opened the pull request:
**What this PR does / why we need it**:
Create new options for Create and Update (through POST/PUT/PATCH).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

Commit:
dry-run: Create new options for Update/Create and pass it along

Commit:
dry-run: Run generate commands for new Options types

lavalamp commented on pkg/registry/apps/deployment/registry.go:
```diff
@@ -65,7 +65,7 @@ func (s *storage) GetDeployment(ctx context.Context, deploymentID string, option
 
 func (s *storage) CreateDeployment(ctx context.Context, deployment *extensions.Deployment, createValidation rest.ValidateObjectFunc) (*extensions.Deployment, error) {
-	obj, err := s.Create(ctx, deployment, createValidation, false)
+	obj, err := s.Create(ctx, deployment, createValidation, &metav1.CreateOptions{})
```
Maybe make a metav1.CreateOptionsTODO() function so you can be sure to find and fix all of these after the fact? If you go this route, anyway.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -453,6 +453,36 @@ type DeleteOptions struct {
+
+	// DryRun decides if the request should be completed or not.
+	// +optional
+	DryRun bool `json:"dryRun,omitempty" protobuf:"varint,1,opt,name=dryRun"`
```
I think the decision from the call yesterday was to make this an enumerated string.

apelisse replied:
Absolutely, I haven't taken the time to update yet.

deads2k commented:
@sttts I'm behind at the moment.  I've got a [comment](https://github.com/kubernetes/kubernetes/pull/65154#discussion_r197108505) on the prereq pull and the API here needs an update.  Got time while I deal with the downstream?

sttts commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -468,6 +468,10 @@ type CreateOptions struct {
+
+	// If IncludeUninitialized is specified, the object may be
+	// returned without completing initialization.
+	IncludeUninitialized bool `json:"includeUninitialized,omitempty" protobuf:"varint,2,opt,name=includeUninitialized"`
```
this surprises me. How is this related to dry-run? Why do we expose that via an API "suddenly"?

sttts replied:
Also note that initializers go away (if I have understood @smarterclayton correctly). We should not expose a parameter for something that is planned to be removed.

sttts replied:
Ok, found it: https://github.com/kubernetes/kubernetes/pull/65105/commits/69e757ae6077d31376512e8b6b3e54fbe90a30f6#diff-7861f33474e4eb8d21cf1d33de5e4d4eL114 had the logic before, with a manually parsed get var. So nvm my comment above.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go:
```diff
@@ -260,6 +260,14 @@ func DeleteCollection(r rest.CollectionDeleter, checkBody bool, scope RequestSco
 				ae := request.AuditEventFrom(ctx)
 				audit.LogRequestObject(ae, obj, scope.Resource, scope.Subresource, scope.Serializer)
+			} else {
+				if values := req.URL.Query(); len(values) > 0 {
```
this is guarded by `if checkBody`. Looks odd.

sttts replied:
In the delete handler we call the guarding var `allowsOptions`.

apelisse replied:
Yeah, I'm not sure what the point of these variable is, but it looks like "checkBody" and  "allowOptions" have the same goal here?

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -111,7 +111,9 @@ func createHandler(r rest.NamedCreater, scope RequestScope, admit admission.Inte
 
 		// TODO: replace with content type negotiation?
-		includeUninitialized := req.URL.Query().Get("includeUninitialized") == "1"
+		options := &metav1.CreateOptions{
```
Shouldn't this be using the parameter decoder? This looks hacky.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -80,6 +81,24 @@ func createHandler(r rest.NamedCreater, scope RequestScope, admit admission.Inte
+				}
+			}
+			// Timeout is not part of the options, drop it.
+			delete(values, "timeout")
```
DecodeParameters ignores params that are not in the struct (necessary for new clients speaking to old servers). So, no need for the delete.

apelisse replied:
Fixed

sttts commented on pkg/master/reconcilers/lease.go:
```diff
@@ -221,7 +221,8 @@ func (r *leaseEndpointReconciler) doReconcile(serviceName string, endpointPorts
 
 	glog.Warningf("Resetting endpoints for master service %q to %v", serviceName, masterIPs)
-	return r.endpointRegistry.UpdateEndpoints(ctx, e, rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc)
+	// PLEASE_PAY_ATTENTION(apelisse): No idea if this is what we need to do here. Does it make sense to have DryRun here anyway?
```
this is the leader election code which speaks directly to the registry to update the endpoint. So this certainly does not need do a dry-run here. It is very very special purpose.

apelisse replied:
Fixed

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:
```diff
@@ -89,6 +90,17 @@ func PatchResource(r rest.Patcher, scope RequestScope, admit admission.Interface
+		options := &metav1.UpdateOptions{}
+		if values := req.URL.Query(); len(values) > 0 {
+			// Timeout is not part of the options, drop it.
+			delete(values, "timeout")
```
same here, DecodeParameters drops unknown fields.

apelisse replied:
Fixed

sttts commented:
Looks generally good. Comments for @jennybuckley's change in her PR.

apelisse commented:
@sttts Feel free to take another look. Still needs to be rebased, but I'm waiting for Jenny's PR to be rebased first.

I've added a couple of clean-up commits that are somewhat unrelated. I've also addressed your comment and changed the dryRun flag to a string.
Thanks

sttts commented on pkg/master/reconcilers/lease.go:
```diff
@@ -221,7 +221,8 @@ func (r *leaseEndpointReconciler) doReconcile(serviceName string, endpointPorts
 
 	glog.Warningf("Resetting endpoints for master service %q to %v", serviceName, masterIPs)
-	return r.endpointRegistry.UpdateEndpoints(ctx, e, rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc)
+	// Use default options, dry-run doesn't make sense here.
```
would even drop the dry-run part. Dry-run=false is the default case. And it's obvious that we want to do a real update there. The sentence confuses more than it helps.

sttts commented on pkg/registry/apps/deployment/storage/storage_test.go:
```diff
@@ -296,7 +296,7 @@ func TestStatusUpdate(t *testing.T) {
 	}
 
-	if _, _, err := storage.Status.Update(ctx, update.Name, rest.DefaultUpdatedObjectInfo(&update), rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc); err != nil {
+	if _, _, err := storage.Status.Update(ctx, update.Name, rest.DefaultUpdatedObjectInfo(&update), rest.DefaultUpdateConfig(rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc, false), &metav1.UpdateOptions{}); err != nil {
```
@jennybuckley I haven't taken another look on your PR, yet. From the changes here: wouldn't a constant like `UpdateNoDryRun` (and a `type UpdateMode bool` type) be much easier to read? This `false` or even a `true` value looks far too innocent.

sttts commented:
@apelisse can you squash your commits? It's a bit hard to review with @jennybuckley's update signature changes all over the place (and Github's inability to show a commit range only).

apelisse commented:
Squashed. Except for the code generation, that I've left in a separate commit.

apelisse commented:
@sttts Rebased and ready for merge!

sttts commented:
/assign @liggitt 

for the registry changes.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -80,6 +81,21 @@ func createHandler(r rest.NamedCreater, scope RequestScope, admit admission.Inte
+		values := req.URL.Query()
+		// TODO: replace with content type negotiation?
+		// "1" is accepted as "true" for backward compatibility.
+		for i, v := range values["includeUninitialized"] {
```
this could be moved into a helper maybe. Also, do we regression test this?

liggitt replied:
Convert_Slice_string_To_bool should already handle this conversion

liggitt commented on pkg/api/testing/serialization_test.go:
```diff
@@ -156,8 +156,10 @@ var nonRoundTrippableTypes = sets.NewString(
+	"UpdateOptions",
 )
 
 var commonKinds = []string{"Status", "ListOptions", "DeleteOptions", "ExportOptions"}
```
do CreateOptions/UpdateOptions need to be added to this list?

apelisse replied:
I don't know, but let's do it for consistency.

sttts replied:
GetOptions

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -418,6 +418,13 @@ const (
+const (
+	// DryRunAll means that the entire chain should be verified: admission
+	// chain, validation, merging, everything.
+	DryRunAll = "All"
```
if this is what we call this, let's make sure we agree that any new stages in the api pipeline in the future will get included in this mode, so that we never ever ever have to have a repeat of the "all does not mean all" issue. cc @smarterclayton

liggitt replied:
I don't think this is referenced anywhere yet. If we want to drop this from this PR to get the signature/plumbing changes in, and continue discussion of the semantics of the parameter, that would be fine with me

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -453,6 +460,36 @@ type DeleteOptions struct {
 	// +optional
 	PropagationPolicy *DeletionPropagation `json:"propagationPolicy,omitempty" protobuf:"varint,4,opt,name=propagationPolicy"`
+
+	// DryRun decides if the request should be completed or not.
```
make the documentation match the serialized field name: `dryRun`

the documentation for this field should:
* indicate the guarantee of setting the field (the mutation will not be persisted)
* indicate what the recognized values mean ("All" means ____)
* indicate what will be done with unrecognized values

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:
```diff
@@ -106,10 +115,30 @@ func PatchResource(r rest.Patcher, scope RequestScope, admit admission.Interface
 
 		userInfo, _ := request.UserFrom(ctx)
-		staticAdmissionAttributes := admission.NewAttributesRecord(nil, nil, scope.Kind, namespace, name, scope.Resource, scope.Subresource, admission.Update, userInfo)
+		staticAdmissionAttributes := admission.NewAttributesRecord(
```
No, it hasn't. Sorry, I initially had multiple commits, and I have a couple of pure cosmetic changes in the middle that got squashed together.

apelisse commented:
/retest

apelisse commented:
@liggitt Updated!

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -110,17 +120,14 @@ func createHandler(r rest.NamedCreater, scope RequestScope, admit admission.Inte
 		}
 
-		// TODO: replace with content type negotiation?
-		includeUninitialized := req.URL.Query().Get("includeUninitialized") == "1"
```
can you verify this got populated/converted properly ("1" and "true" set to true, unset or anything else set to false)?

liggitt commented:
/lgtm

deads2k commented:
/approve

liggitt commented:
/hold

forgot we didn't settle the param type... string vs []string

deads2k commented:
> forgot we didn't settle the param type... string vs []string

Heh, I just thought to ask about that too.

/approve cancel

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -453,6 +453,42 @@ type DeleteOptions struct {
 	// +optional
 	PropagationPolicy *DeletionPropagation `json:"propagationPolicy,omitempty" protobuf:"varint,4,opt,name=propagationPolicy"`
+
+	// Whether the request should be completed and modifications will not be persisted.
```
I'd suggest something more like this:

> When present, indicates that modifications should not be persisted.
> An invalid or unrecognized dryRun directive will result in a BadRequest response and no further processing of the request.

and then we can add details about what directives are supported as we determine that

apelisse commented:
/retest

deads2k commented:
type looks correct now.

/approve

liggitt commented:
/hold cancel
/lgtm

apelisse commented:
/retest

roycaihw commented:
It looks like we added `meta.v1.CreateOptions` and `meta.v1.UpdateOptions` in this PR, but they weren't reflected in our openapi spec. The `dryRun` query parameter was only added to the delete operations in the spec. I'm wondering if this was intended?
